circuit/src/gadgets/main.rs

```diff
@@ -0,0 +1,950 @@
+		let instance = meta.instance_column();
+
+		advice.map(|c| meta.enable_equality(c));
+		fixed.map(|c| meta.enable_constant(c));
+		meta.enable_equality(instance);
+
+		let selector = meta.selector();
+
+		Self { advice, fixed, instance, selector }
+	}
+
+	/// Temporary construct for testing
+	pub fn construct_from_selector<F: FieldExt>(common: &CommonConfig, selector: Selector) -> Self {
+		let advice = common.advice.split_at(NUM_ADVICE).0.try_into().unwrap();
+		let fixed = common.fixed.split_at(NUM_FIXED).0.try_into().unwrap();
+		Self { advice, fixed, instance: common.instance.clone(), selector: selector.clone() }
+	}
+}
+
+/// Structure for the main chip.
+pub struct MainChip<F: FieldExt> {
+	advice: [AssignedCell<F, F>; NUM_ADVICE],
+	fixed: [F; NUM_FIXED],
+}
+
+impl<F: FieldExt> MainChip<F> {
+	/// Assigns a new witness that is equal to boolean AND of `x` and `y`
+	pub fn new(advice: [AssignedCell<F, F>; NUM_ADVICE], fixed: [F; NUM_FIXED]) -> Self {
+		Self { advice, fixed }
+	}
+}
+
+impl<F: FieldExt> Chip<F> for MainChip<F> {
+	type Output = ();
+
+	fn configure(
+		common: &crate::CommonConfig, meta: &mut ConstraintSystem<F>,
+	) -> halo2::plonk::Selector {
+		let selector = meta.selector();
+
+		meta.create_gate("main gate", |v_cells| {
+			// MainGate constraints
+			let a = v_cells.query_advice(common.advice[0], Rotation::cur());
+			let b = v_cells.query_advice(common.advice[1], Rotation::cur());
+			let c = v_cells.query_advice(common.advice[2], Rotation::cur());
+			let d = v_cells.query_advice(common.advice[3], Rotation::cur());
+			let e = v_cells.query_advice(common.advice[4], Rotation::cur());
+
+			let sa = v_cells.query_fixed(common.fixed[0], Rotation::cur());
+			let sb = v_cells.query_fixed(common.fixed[1], Rotation::cur());
+			let sc = v_cells.query_fixed(common.fixed[2], Rotation::cur());
+			let sd = v_cells.query_fixed(common.fixed[3], Rotation::cur());
+			let se = v_cells.query_fixed(common.fixed[4], Rotation::cur());
+
+			let s_mul_ab = v_cells.query_fixed(common.fixed[5], Rotation::cur());
+			let s_mul_cd = v_cells.query_fixed(common.fixed[6], Rotation::cur());
+
+			let s_constant = v_cells.query_fixed(common.fixed[7], Rotation::cur());
+
+			let selector = v_cells.query_selector(selector);
+
+			vec![
+				selector
+					* (a.clone() * sa
+						+ b.clone() * sb + c.clone() * sc
+						+ d.clone() * sd + e * se
+						+ a * b * s_mul_ab + c * d * s_mul_cd
+						+ s_constant),
+			]
+		});
+		selector
+	}
+
+	fn synthesize(
+		self, common: &crate::CommonConfig, selector: &halo2::plonk::Selector,
+		mut layouter: impl Layouter<F>,
+	) -> Result<Self::Output, Error> {
+		layouter.assign_region(
+			|| "main gate",
+			|region| {
+				let mut ctx = RegionCtx::new(region, 0);
+
+				ctx.enable(*selector)?;
+
+				let result = self
+					.advice
+					.clone()
+					.into_iter()
+					.enumerate()
+					.map(|(i, v)| ctx.assign_advice(common.advice[i], v.value().cloned()))
```

Comment:
This should be copy constrant, not a regular assignment

Reply:
Done in https://github.com/eigen-trust/protocol/pull/132/commits/7fcb9525b5ce9a8ec1d6969b7ca3a616475b1a06